source/simple_sub_pub_demo.c

```diff
@@ -512,19 +512,27 @@ static void prvSimpleSubscribePublishTask( void * pvParameters )
         /* The value by the callback that executed when the publish was acked
          * came from the context passed into MQTTAgent_Publish() above, so
          * should match the value set in the context above. */
-        configASSERT( ulNotification == ulValueToNotify );
+        configASSERT( ( ulNotification == ulValueToNotify ) || ( xQoS == 0 ) );
 
         if( ulNotification == ulValueToNotify )
         {
-            LogInfo( ( "Received ack from publishing to topic %s. Sleeping for %d ms.",
-                       pcTopicBuffer,
-                       mqttexampleDELAY_BETWEEN_PUBLISH_OPERATIONS_MS ) );
+            LogInfo( ( "Received ack from publishing to topic %s.",
+                       pcTopicBuffer ) );
         }
         else
         {
-            LogInfo( ( "Error - Timed out or didn't receive ack from publishing to topic %s Sleeping for %d ms.",
-                       pcTopicBuffer,
-                       mqttexampleDELAY_BETWEEN_PUBLISH_OPERATIONS_MS ) );
+        	if( xQoS == 0 )
+        	{
+        		ulQoS0FailureCount++;
+        	}
+        	else
```

Comment:
Since there's a config assert for `( ulNotification == ulValueToNotify ) || ( xQoS == 0 )` will this `else` be hit without triggering the assert, as `xQoS != 0`? Should the assert just be removed?

Reply:
I think the || makes this ok.  It is saying it is ok for ulNotification to not equal ulValueToNotify if xQoS was 0 because QoS 0 has no guaranteed delivery (therefore it is not necessarily an error if a message is missing).  So if ulNotifycation == ulValueToNotify evaluates to false but xQoS == 0 evaluates to true then the expression ( false || true ) will evaluate to true and the assert won't be triggered.  Will add a comment into the code.

Reply:
So the code without logs looks like this:
```
        configASSERT( ( ulNotification == ulValueToNotify ) || ( xQoS == 0 ) );

        if( ulNotification == ulValueToNotify )
        {
        }
        else
        {
        	if( xQoS == 0 )
        	{
        		ulQoS0FailureCount++;
        	}
        	else
        	{
        		ulQoS1FailureCount++;    // <- How will we hit this without triggering the assert?
        	}
        }
```
I don't see how ulQoS1FailureCount will be incremented without first triggering the assert. And, if we're tracking the number of failures, should we just remove the assert to begin with?

Also, the notification we get isn't from receiving a publish (as `prvIncomingPublishCallback` doesn't send any notifications, only logs the incoming message), but from when our publish command has completed, so we should get a notification for the QoS 0 case as soon as we send the packet. If we're not getting that notification for QoS 0, that would indicate the MQTT agent hasn't processed our command.

Reply:
Intent is to move some asserts out as they stop the soak test, sometimes unnecessarily, and they also prevent testing the disconnect and reconnect behaviour as disconnecting triggers the asserts.  I will have another pass at this but prefer to do so after checking something in to make sure the branch doesn't get out of date.  For now I may just move that code so we can merge, then create a separate PR with the assert behaviour changed.